### PR TITLE
Provide Array.fill in common stdlib

### PR DESCRIPTION
**Pull request by JakeWharton**
KT-32781

### CODE REVIEW

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt [hunk 1]
```diff
@@ -1273,17 +1273,40 @@ object ArrayOps : TemplateGroupBase() {
...
 
+        on(Platform.JVM) {
+            suppress("ACTUAL_FUNCTION_WITH_DEFAULT_ARGUMENTS")
+            body {
+                """
+                java.util.Arrays.fill(this, fromIndex, toIndex, element)
+                """
+            }
+        }
+        on(Platform.JS) {
```
Comment (JakeWharton) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: JS has https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill but not sure it's worth delegating to or not.

File: libraries/stdlib/common/src/generated/_Arrays.kt [hunk 2]
```diff
@@ -6338,6 +6338,51 @@ public expect fun BooleanArray.copyOfRange(fromIndex: Int, toIndex: Int): Boolea
  */
 public expect fun CharArray.copyOfRange(fromIndex: Int, toIndex: Int): CharArray
 
+/**
+ * Fills original array with the provided value.
+ */
+public expect fun <T> Array<T>.fill(element: T, fromIndex: Int = 0, toIndex: Int = size): Unit
```
Comment (Prototik) [inline libraries/stdlib/common/src/generated/_Arrays.kt]: Isn't `Array<in T>` enough here?

Comment (JakeWharton) [reply libraries/stdlib/common/src/generated/_Arrays.kt]: This is the original signature of the function. The PR just moves it to common.

If we want a signature change I would prefer to do it separately.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 1 above)
Comment (ilya-g) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: I think it's definitely worth it, however we should still check argument values to enforce our contract of this method.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt [hunk 3]
```diff
@@ -1273,17 +1273,40 @@ object ArrayOps : TemplateGroupBase() {
...
+            if (fromIndex !in 0..size) {
+              throw IndexOutOfBoundsException("fromIndex ${'$'}fromIndex out of range [0, ${'$'}size]")
+            }
+            if (toIndex > size) {
+              throw IndexOutOfBoundsException("toIndex ${'$'}toIndex out of range [${'$'}fromIndex, ${'$'}size]")
+            }
+            require(fromIndex <= toIndex) { "fromIndex(${'$'}fromIndex) > toIndex(${'$'}toIndex)" }
+            for (index in fromIndex until toIndex) {
+                this[index] = element
+            }
```
Comment (ilya-g) [inline libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Given that Kotlin/Native has its own implementation and Kotlin/JS will be delegating to array's method, there seem to be no need in this manual implementation.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 1 above)
Comment (JakeWharton) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Ok.

 > we should still check argument values to enforce our contract of this method

Yes JS seems to just silently ignore out-of-bounds and to < from 😑

> We'll need to provide polyfills though, as we target ES 5 that doesn't have these methods in the standard.

Do you know of a similar API does this that I can copy? I can go searching for one, but thought you might know one off-hand.

File: libraries/stdlib/common/src/generated/_Arrays.kt (hunk 2 above)
Comment (ilya-g) [reply libraries/stdlib/common/src/generated/_Arrays.kt]: @Prototik, the current signature won't benefit from changing `Array<T>` to `Array<in T>`.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 3 above)
Comment (JakeWharton) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Oh, ok. Do I point to its implementation somehow? Or remove it from being generated? I don't really understand how native tracks these kind of changes being in a separate repo.

Comment (ilya-g) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Yes, you need to call implementation methods (e.g. https://github.com/JetBrains/kotlin-native/blob/9e26d32/runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt#L78) here, however building Kotlin/Native together with Kotlin can be quite complicated, so feel free to leave this exercise to maintainers :-)

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 1 above)
Comment (ilya-g) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: The MDN article you have linked contains a polyfill implementation. We keep 'em collection of polyfills used in stdlib here: https://github.com/JetBrains/kotlin/blob/1.3.50/libraries/stdlib/js-v1/src/js/polyfills.js

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 3 above)
Comment (JakeWharton) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Ok I left it as an `arrayFill` function call, but obviously implementations will need to be added to that file similar to `arrayCopy`.

It's not clear to me whether the responsibility for the argument checks should live in the template for native or in the implementation. If it's in the template, let me know and I'll update.

File: libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt (hunk 1 above)
Comment (JakeWharton) [reply libraries/tools/kotlin-stdlib-gen/src/templates/Arrays.kt]: Updated. PTAL!

**Comment by ilya-g** [comment]
Thanks, I've merged it manually with several edits, see the commit range https://github.com/JetBrains/kotlin/compare/1a6069382ecafc4c6d79bc109c83f67f0aca6188~1...2e445ebad6b95bee38749d85274abe3742baae68

**Comment by JakeWharton** [comment]
Thanks! I hope these PRs help more than they just obligate you to do more work.

**Pull request closed without merging**